extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java

```diff
@@ -117,8 +121,21 @@ public AgroalDataSource createDataSource(String dataSourceName,
                     .principal(new NamePrincipal(dataSourceRuntimeConfig.username.get()));
         }
         if (dataSourceRuntimeConfig.password.isPresent()) {
+
+            String passwordPropertyKey = "quarkus.datasource"
+                    + (dataSourceName.equals("<default>") ? "" : "." + dataSourceName)
+                    + ".password";
+
             agroalConnectionFactoryConfigurationSupplier
-                    .credential(new SimplePassword(dataSourceRuntimeConfig.password.get()));
+                    .credential(new ConfigPropertyPassword(passwordPropertyKey));
+        }
+
+        // Vault Authentication
+        VaultDataSourceRuntimeConfig vaultDataSourceRuntimeConfig = dataSourceRuntimeConfig.vault;
+        if ((vaultDataSourceRuntimeConfig.dbrole.isPresent())) {
+            VaultService vaultService = Arc.container().instance(VaultService.class).get(); // TODO is this the right way?
```

Comment:
Since a producer is available, `VaultService` should be injected ?

Reply:
good point. done